capsules/src/test/aes.rs

```diff
@@ -0,0 +1,214 @@
+//! Test the AES hardware
+
+use kernel::ReturnCode;
+use kernel::hil;
+use kernel::hil::symmetric_encryption::{AES128_BLOCK_SIZE, AES128, AES128Ctr, AES128CBC};
+use kernel::common::take_cell::TakeCell;
+use core::cell::Cell;
+
+pub struct Test<'a, A: 'a> {
+    aes: &'a A,
+
+    key: TakeCell<'a, [u8]>,
+    iv: TakeCell<'a, [u8]>,
+    source: TakeCell<'a, [u8]>,
+    data: TakeCell<'a, [u8]>,
+
+    encrypting: Cell<bool>,
+    mode_ctr: Cell<bool>,
+    use_source: Cell<bool>,
+}
+
+pub const DATA_OFFSET: usize = AES128_BLOCK_SIZE;
+const DATA_LEN: usize = 4 * AES128_BLOCK_SIZE;
+
+impl<'a, A: AES128<'a> + AES128Ctr + AES128CBC> Test<'a, A> {
+    pub fn new(aes: &'a A,
+               key: &'a mut [u8],
+               iv: &'a mut [u8],
+               source: &'a mut [u8],
+               data: &'a mut [u8]) -> Self {
+        Test {
+            aes: aes,
+
+            key: TakeCell::new(key),
+            iv: TakeCell::new(iv),
+            source: TakeCell::new(source),
+            data: TakeCell::new(data),
+
+            encrypting: Cell::new(true),
+            mode_ctr: Cell::new(true),
+            use_source: Cell::new(true),
+        }
+    }
+
+    pub fn run(&self) {
+        self.aes.enable();
+
+        // Copy key into key buffer and configure it in the hardware
+        self.key.map(|key| {
+            for (i,b) in KEY.iter().enumerate() {
+                key[i] = *b;
+            }
+
+            assert!(self.aes.set_key(key) == ReturnCode::SUCCESS);
+        });
```

Comment:
well, I think it is better to use the following syntax:

```rust
let return = self.key.map(|key| {
  // mutate you buffer
  self.aes.set_key(key)
}).unwrap_or_else(|| ReturnCode::WhatYouWant);
assert!(return == ReturnCode::SUCCESS)
```
Because if `your` TakeCell is `None` the closure will not be executed but in this case you are probably sure
that is not None